### PR TITLE
Path planning

### LINKED ISSUE #10: Get path planning working
There are a couple of different path planning options we can go with:

1. https://docs.wpilib.org/en/latest/docs/software/examples-tutorials/trajectory-tutorial/index.html
2. https://docs.wpilib.org/en/latest/docs/software/wpilib-tools/pathfinder/index.html
3. https://docs.wpilib.org/en/latest/docs/software/advanced-control/trajectories/index.html

so we need to figure out which method works best for us and then implement it.

### PR DESCRIPTION
**Pull request by BethanNguyen**
resolves #10 

Changes made by bethan and en

Path Planning is working:
 - DriveTrain has autonomous methods
  - getPoseListFromPathWeaverJson and getAutonomousCommandFromTrajectory
  - tankDriveVelocity
 - 2 Autos
  - AutoBrettV7
  - TestAuto

### CODE REVIEW

File: src/main/java/frc/robot/Constants.java [hunk 1]
```diff
@@ -26,6 +30,37 @@
...
+
+    // If Measured in meters trackWidth should be 0.58m
+    public static final double trackWidth = 0.5792883856032899;
+    public static final DifferentialDriveKinematics kinematics = new DifferentialDriveKinematics(
+        trackWidth);
+
+    public static final DifferentialDriveVoltageConstraint VOLTAGE_CONSTRAINT =
+        new DifferentialDriveVoltageConstraint(feedForward, kinematics, maxVoltageAuto);
+
+    // Keep in case we want to create trajectory without PathWeaver
```
Comment (BethanNguyen) [inline src/main/java/frc/robot/Constants.java]: heck this comment

File: src/main/java/frc/robot/RobotContainer.java [hunk 2]
```diff
@@ -35,20 +39,28 @@
...
   private XboxController driveController;
   private XboxController altController;
-  private CommandBase auto;
+
   private final DriveTrain drive;
   private final Intake intake;
 
+  private SendableChooser<Command> autoChooser;
+
+
```
Comment (BethanNguyen) [inline src/main/java/frc/robot/RobotContainer.java]: heck this blank line

File: src/main/java/frc/robot/RobotContainer.java [hunk 3]
```diff
@@ -58,8 +70,11 @@ public RobotContainer() {
                     driveController.getTriggerAxis(GenericHID.Hand.kLeft),
                     driveController.getX(GenericHID.Hand.kLeft)),
             drive));
+
+
```
Comment (BethanNguyen) [inline src/main/java/frc/robot/RobotContainer.java]: bad juice

File: src/main/java/frc/robot/commands/ZeroEncoder.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
+package frc.robot.commands;
+
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import frc.robot.subsystems.DriveTrain;
+
+public class ZeroEncoder extends CommandBase {
+
+  DriveTrain drive;
```
Comment (BethanNguyen) [inline src/main/java/frc/robot/commands/ZeroEncoder.java]: Needs to be private

File: src/main/java/frc/robot/commands/ZeroEncoder.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
+package frc.robot.commands;
```
Comment (BethanNguyen) [inline src/main/java/frc/robot/commands/ZeroEncoder.java]: move commands into different packages by their subsystem

File: src/main/java/frc/robot/commands/auto/AutoBrettV7.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+import edu.wpi.first.wpilibj2.command.SequentialCommandGroup;
+import edu.wpi.first.wpilibj2.command.WaitCommand;
+import frc.robot.subsystems.DriveTrain;
+
+/**
+ * Drives Forward 1.5 meters, waits 3 seconds and travels another 1.5 meters.
+ */
+public class AutoBrettV7 extends SequentialCommandGroup {
+
+  DriveTrain drive;
```
Comment (BethanNguyen) [inline src/main/java/frc/robot/commands/auto/AutoBrettV7.java]: needs to be private

File: src/main/java/frc/robot/commands/auto/TestReverse.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+import edu.wpi.first.wpilibj.geometry.Pose2d;
+import edu.wpi.first.wpilibj.trajectory.Trajectory;
+import edu.wpi.first.wpilibj.trajectory.TrajectoryGenerator;
+import edu.wpi.first.wpilibj2.command.InstantCommand;
+import edu.wpi.first.wpilibj2.command.SequentialCommandGroup;
+import edu.wpi.first.wpilibj2.command.WaitCommand;
+import frc.robot.subsystems.DriveTrain;
+
+public class TestReverse extends SequentialCommandGroup {
```
Comment (BethanNguyen) [inline src/main/java/frc/robot/commands/auto/TestReverse.java:13]: Comment describing auto

File: src/main/java/frc/robot/commands/auto/TestReverse.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+  private DriveTrain drive;
+
+  private Trajectory trajectoryForward;
+  private Trajectory trajectoryReverse;
+
+  private Pose2d pose;
+
+  /**
+   * Auto command for testing.
```
Comment (BethanNguyen) [inline src/main/java/frc/robot/commands/auto/TestReverse.java]: comment on class should cover this, so delete

File: src/main/java/frc/robot/commands/auto/TestReverse.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+import frc.robot.subsystems.DriveTrain;
+
+public class TestReverse extends SequentialCommandGroup {
+
+  private DriveTrain drive;
+
+  private Trajectory trajectoryForward;
+  private Trajectory trajectoryReverse;
+
+  private Pose2d pose;
```
Comment (BethanNguyen) [inline src/main/java/frc/robot/commands/auto/TestReverse.java]: not needed

File: src/main/java/frc/robot/RobotContainer.java [hunk 10]
```diff
@@ -70,20 +85,24 @@ private void configureButtonBindings() {
     driveController = new XboxController(driveControllerPort);
     altController = new XboxController(altControllerPort);
 
+    // Runs Intake
```
Comment (BethanNguyen) [inline src/main/java/frc/robot/RobotContainer.java]: bad

File: src/main/java/frc/robot/RobotContainer.java [hunk 11]
```diff
@@ -58,8 +70,11 @@ public RobotContainer() {
                     driveController.getTriggerAxis(GenericHID.Hand.kLeft),
                     driveController.getX(GenericHID.Hand.kLeft)),
             drive));
+
+
   }
 
+
```
Comment (BethanNguyen) [inline src/main/java/frc/robot/RobotContainer.java]: delete blank line

File: src/main/java/frc/robot/subsystems/DriveTrain.java [hunk 12]
```diff
@@ -7,26 +7,58 @@
...
+import java.io.InputStream;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import javax.json.Json;
+import javax.json.JsonArray;
+import javax.json.JsonObject;
+import javax.json.JsonReader;
 
+@SuppressWarnings("FieldCanBeLocal")
```
Comment (BethanNguyen) [inline src/main/java/frc/robot/subsystems/DriveTrain.java:58]: delete

File: src/main/java/frc/robot/subsystems/DriveTrain.java [hunk 13]
```diff
@@ -36,29 +68,76 @@
   private WPI_VictorSPX leftFollower1;
   private WPI_VictorSPX leftFollower2;
 
+  private DifferentialDrive drive;
+
+  private AHRS gyro;
+  private double gyroInversionNumber;
```
Comment (BethanNguyen) [inline src/main/java/frc/robot/subsystems/DriveTrain.java]: delete

File: src/main/java/frc/robot/subsystems/DriveTrain.java [hunk 14]
```diff
@@ -36,29 +68,76 @@
...
 
     leftFollower1.follow(leftMaster);
     leftFollower2.follow(leftMaster);
+
+    drive = new DifferentialDrive(leftMaster, rightMaster);
+    drive.setSafetyEnabled(false);
+    drive.setRightSideInverted(false);
+
+    gyro = new AHRS();
+    gyroInversionNumber = -1;
```
Comment (BethanNguyen) [inline src/main/java/frc/robot/subsystems/DriveTrain.java]: delete

File: src/main/java/frc/robot/subsystems/DriveTrain.java [hunk 15]
```diff
@@ -129,4 +232,195 @@ public double getYaw() {
...
   }
-}
+
+  public void zeroYaw() {
+    gyro.zeroYaw();
+  }
+
+  public void setGyroInverted(boolean inversion) {
+    gyroInversionNumber = (inversion ? -1 : 1);
+  }
```
Comment (BethanNguyen) [inline src/main/java/frc/robot/subsystems/DriveTrain.java]: delete

File: src/main/java/frc/robot/subsystems/DriveTrain.java [hunk 16]
```diff
@@ -129,4 +232,195 @@ public double getYaw() {
...
+    } else {
+      return new TrajectoryConfig(
+          maxVelocityMetersPerSecond,
+          maxAccelerationMetersPerSecondSquared)
+          .setKinematics(kinematics)
+          .setStartVelocity(0)
+          .setEndVelocity(0)
+          .setReversed(false);
+    }
+  }
```
Comment (BethanNguyen) [inline src/main/java/frc/robot/subsystems/DriveTrain.java]: `return new TrajectoryConfig(
          maxVelocityMetersPerSecond,
          maxAccelerationMetersPerSecondSquared)
          .setKinematics(kinematics)
          .setStartVelocity(0)
          .setEndVelocity(0)
          .setReversed(isReversed);`

File: src/main/java/frc/robot/subsystems/DriveTrain.java [hunk 17]
```diff
@@ -129,4 +232,195 @@ public double getYaw() {
...
+   * Replicates data from Pathweaver produced JSON file so that we can input our own Trajectory
+   * Configuration.
+   *
+   * @param trajectoryName Name of Trajectory
+   * @return List of Pose2d objects
+   */
+  public List<Pose2d> getPoseListFromPathWeaverJson(String trajectoryName) {
+    ArrayList<Pose2d> poseList = new ArrayList<>();
+    double x;
+    double y;
```
Comment (BethanNguyen) [inline src/main/java/frc/robot/subsystems/DriveTrain.java]: move to smallest scope

**Review by SHS-Programmers — APPROVED** [review]

**Pull request merged**